core/market/decentralized/src/protocol/discovery/builder.rs

```diff
@@ -0,0 +1,172 @@
+use std::any::{Any, TypeId};
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use crate::protocol::callbacks::{CallbackFuture, OutputFuture};
+use crate::protocol::discovery::DiscoveryImpl;
+use crate::protocol::{CallbackHandler, CallbackMessage, Discovery, HandlerSlot};
+
+#[derive(Default)]
+pub struct DiscoveryBuilder {
+    data: HashMap<TypeId, Box<dyn Any>>,
+    handlers: HashMap<TypeId, Box<dyn Any>>,
+}
+
+impl DiscoveryBuilder {
+    pub fn data<T: Clone + Send + Sync + 'static>(mut self, data: T) -> Self {
+        self.data.insert(TypeId::of::<T>(), Box::new(data));
+        self
+    }
+
+    pub fn add_data_handler<M: CallbackMessage, T: Clone + Send + Sync + 'static>(
+        mut self,
+        mut f: impl DataCallbackHandler<M, T>,
+    ) -> Self {
+        let boxed = self
+            .data
+            .get(&TypeId::of::<T>())
+            .expect("data handler needs data");
```

Comment:
Because this only can happen when the Market developer (one of us) is misusing the builder (forgotten to specify data). It is not user-facing error. it should fail fast during initial testing or demo

Reply:
And the same happens if you return error from initialization

Reply:
Nope, in intialisation you are also handling GSB errors... it might be user-facing

Reply:
Ok, if you insists ;)